cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/SupportBundleHttpHandler.java

```diff
@@ -0,0 +1,552 @@
+                generateApplicationFile(namespaceId, app.getName(), appFolderPath.toString());
+            Map<ProgramRunId, RunRecordDetail> runMap =
+                store.getRuns(
+                    applicationPair.getKey(),
+                    ProgramRunStatus.ALL,
+                    Integer.MAX_VALUE,
+                    meta -> true);
+            long startTs = 0;
+            // Gets the latest run info
+            for (ProgramRunId programRunId : runMap.keySet()) {
+              if (startTs < runMap.get(programRunId).getStartTs()
+                  && programRunId.getProgram().equals("DataPipelineWorkflow")) {
+                latestProgramRunId = programRunId;
+                latestRunRecordDetail = runMap.get(programRunId);
+                startTs = runMap.get(programRunId).getStartTs();
+              }
+            }
+            if (latestProgramRunId != null) {
+              JsonObject metrics =
+                  queryMetrics(
+                      namespaceId,
+                      app.getName(),
+                      latestProgramRunId.getRun(),
+                      workflowName,
+                      applicationPair.getValue().getConfiguration(),
+                      latestRunRecordDetail != null ? latestRunRecordDetail.getStartTs() : 0,
+                      latestRunRecordDetail != null && latestRunRecordDetail.getStopTs() != null
+                          ? latestRunRecordDetail.getStopTs()
+                          : DateTime.now().getMillis());
+              generateLogFileAndRunInfo(
+                  latestProgramRunId, latestRunRecordDetail, appFolderPath.toString(), metrics);
+            }
+          }
+        } else if (appId != null) {
+          File appFolderPath = new File(basePath.toString(), appId);
+          if (!appFolderPath.exists()) {
+            appFolderPath.mkdirs();
+          }
+          Pair<ApplicationId, ApplicationDetail> applicationPair =
+              generateApplicationFile(namespaceId, appId, appFolderPath.toString());
+          latestProgramRunId =
+              new ProgramRunId(namespaceId, appId, ProgramType.WORKFLOW, workflowName, runId);
+          latestRunRecordDetail = store.getRun(latestProgramRunId);
+          JsonObject metrics =
+              queryMetrics(
+                  namespaceId,
+                  appId,
+                  runId,
+                  workflowName,
+                  applicationPair.getValue().getConfiguration(),
+                  latestRunRecordDetail != null ? latestRunRecordDetail.getStartTs() : 0,
+                  latestRunRecordDetail != null && latestRunRecordDetail.getStopTs() != null
+                      ? latestRunRecordDetail.getStopTs()
+                      : DateTime.now().getMillis());
+          generateLogFileAndRunInfo(
+              latestProgramRunId, latestRunRecordDetail, appFolderPath.toString(), metrics);
+        }
+      }
+    } catch (Exception e) {
+      LOG.error("", e);
```

Comment:
Put some error message. Also separate tasks failure should be handled gracefully. Single task failure should not fail all collection. You need to have it separated by separate tasks with proper error handling and potentially retries

Reply:
Tasks should be done concurrently using executor service

Reply:
Yeah right now it's handled gracefully for each task, even though one task fail, others will keep running